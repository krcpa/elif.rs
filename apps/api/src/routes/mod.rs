use axum::Router;

pub fn router() -> Router {
    Router::new()
    // API routes will be added here by the framework's code generation
    // This serves as the base router for application-specific routes
}