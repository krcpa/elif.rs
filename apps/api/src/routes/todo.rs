use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::Json,
    routing::{get, post, patch, delete},
    Router,
};
use serde_json::Value;
use uuid::Uuid;

pub fn router() -> Router {
    Router::new()
        .route("/", post(create_Todo))
        .route("/", get(list_Todo))
        .route("/:id", get(get_Todo))
        .route("/:id", patch(update_Todo))
        .route("/:id", delete(delete_Todo))
}

// <<<ELIF:BEGIN agent-editable:create_Todo>>>
async fn create_Todo(
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    // TODO: Implement create logic
    Ok(Json(serde_json::json!({"id": "placeholder"})))
}
// <<<ELIF:END agent-editable:create_Todo>>>

// <<<ELIF:BEGIN agent-editable:list_Todo>>>
async fn list_Todo() -> Result<Json<Value>, StatusCode> {
    // TODO: Implement list logic
    Ok(Json(serde_json::json!({"items": [], "next": null})))
}
// <<<ELIF:END agent-editable:list_Todo>>>

// <<<ELIF:BEGIN agent-editable:get_Todo>>>
async fn get_Todo(
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, StatusCode> {
    // TODO: Implement get logic
    Ok(Json(serde_json::json!({"id": id})))
}
// <<<ELIF:END agent-editable:get_Todo>>>

// <<<ELIF:BEGIN agent-editable:update_Todo>>>
async fn update_Todo(
    Path(id): Path<Uuid>,
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    // TODO: Implement update logic
    Ok(Json(serde_json::json!({"id": id})))
}
// <<<ELIF:END agent-editable:update_Todo>>>

// <<<ELIF:BEGIN agent-editable:delete_Todo>>>
async fn delete_Todo(
    Path(id): Path<Uuid>,
) -> Result<StatusCode, StatusCode> {
    // TODO: Implement delete logic
    Ok(StatusCode::NO_CONTENT)
}
// <<<ELIF:END agent-editable:delete_Todo>>>
