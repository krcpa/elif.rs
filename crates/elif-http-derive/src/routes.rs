//! Route registration and resource macros
//! 
//! Provides #[routes] and #[resource] macros for route organization.

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ItemImpl, ItemFn, ImplItem, LitStr};

use crate::utils::{extract_http_method_info, extract_resource_info};

/// Route registration macro for impl blocks
/// 
/// Generates route registration code from an impl block containing route methods.
/// This macro scans all methods in the impl block and generates a router setup function.
/// 
/// Example:
/// ```rust,ignore
/// #[routes]
/// impl ApiRoutes {
///     #[get("/health")]
///     fn health() -> HttpResult<ElifResponse> {
///         Ok(response().json(json!({"status": "ok"})))
///     }
///     
///     #[resource("/users")]
///     fn users() -> UserController { UserController::new() }
/// }
/// ```
pub fn routes_impl(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input_impl = parse_macro_input!(input as ItemImpl);
    let impl_name = if let syn::Type::Path(type_path) = &*input_impl.self_ty {
        if let Some(segment) = type_path.path.segments.last() {
            &segment.ident
        } else {
            return syn::Error::new_spanned(
                &input_impl.self_ty,
                "Cannot get identifier from an empty type path. Hint: Use a proper struct name like `impl MyRoutes`.",
            )
            .to_compile_error()
            .into();
        }
    } else {
        return syn::Error::new_spanned(
            &input_impl.self_ty, 
            "Expected a simple type for the impl block. Hint: Apply #[routes] to `impl MyStruct { ... }` not complex types."
        )
        .to_compile_error()
        .into();
    };
    
    let mut route_registrations = Vec::new();
    let mut resource_registrations = Vec::new();
    
    // Process each method in the impl block
    for item in &input_impl.items {
        if let ImplItem::Fn(method) = item {
            let method_name = &method.sig.ident;
            
            // Check for HTTP method attributes
            if let Some((http_method, path)) = extract_http_method_info(&method.attrs) {
                let path = if path.is_empty() { "/" } else { &path };
                route_registrations.push(quote! {
                    router = router.#http_method(#path, Self::#method_name);
                });
            }
            
            // Check for resource attribute
            if let Some(resource_path) = extract_resource_info(&method.attrs) {
                resource_registrations.push(quote! {
                    router = router.resource(#resource_path, Self::#method_name());
                });
            }
        }
    }
    
    // Count routes and resources
    let route_count = route_registrations.len();
    let resource_count = resource_registrations.len();
    
    // Generate the router build function
    let expanded = quote! {
        #input_impl
        
        impl #impl_name {
            /// Generated router setup function  
            pub fn build_router() -> String {
                // In production, this would return a proper router instance
                // For now, return a string for testing purposes
                format!("Router with {} routes and {} resources", #route_count, #resource_count)
            }
        }
    };
    
    TokenStream::from(expanded)
}

/// Resource macro for automatic RESTful resource registration
/// 
/// Registers a controller as a RESTful resource with standard routes.
/// 
/// Example:
/// ```rust,ignore
/// #[resource("/users")]
/// fn user_routes() -> UserController { UserController::new() }
/// ```
pub fn resource_impl(args: TokenStream, input: TokenStream) -> TokenStream {
    let path_lit = parse_macro_input!(args as LitStr);
    let input_fn = parse_macro_input!(input as ItemFn);
    
    let resource_path = path_lit.value();
    
    let func_name = &input_fn.sig.ident;
    let resource_path_fn = quote::format_ident!("{}_resource_path", func_name);
    
    // Add resource attribute metadata to the function
    let expanded = quote! {
        #[doc = concat!("RESTful resource at: ", #resource_path)]
        #input_fn
        
        // Generate a helper function with resource metadata
        #[allow(non_snake_case)]
        pub fn #resource_path_fn() -> &'static str {
            #resource_path
        }
    };
    
    TokenStream::from(expanded)
}