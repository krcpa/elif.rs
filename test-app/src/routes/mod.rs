use axum::Router;

pub fn router() -> Router {
    Router::new()
        // Routes will be added here by `elif route add` command
        // Example: .route("/hello", get(crate::controllers::hello_controller))
}
