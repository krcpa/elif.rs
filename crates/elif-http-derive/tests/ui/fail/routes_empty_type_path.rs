//! Test that #[routes] macro fails gracefully with empty type paths

use elif_http_derive::routes;

// This should fail with a proper error message
#[routes]
impl <> {
    pub fn dummy() {}
}

fn main() {}