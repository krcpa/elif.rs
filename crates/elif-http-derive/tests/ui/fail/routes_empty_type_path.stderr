error: expected type, found `{`
 --> tests/ui/fail/routes_empty_type_path.rs:7:9
  |
7 | impl <> {
  |         ^ expected type
